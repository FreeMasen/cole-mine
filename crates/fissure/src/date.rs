use core::fmt;
use std::str::FromStr;

use serde::{Deserialize, Serialize};
use structsy::derive::{embedded_queries, PersistentEmbedded};
use time::{OffsetDateTime, PrimitiveDateTime};

use crate::Result;

#[derive(Debug, PartialEq, Eq, Clone, Copy, PersistentEmbedded, bon::Builder)]
pub struct DateTime {
    pub year: u16,
    pub month: u8,
    pub day: u8,
    #[builder(default)]
    pub hour: u8,
    #[builder(default)]
    pub minute: u8,
    #[builder(default)]
    pub second: u8,
}

#[embedded_queries(DateTime)]
pub trait DateTimeQuery {
    fn with_year(self, year: u16) -> Self;
    fn with_month(self, month: u8) -> Self;
    fn with_day(self, day: u8) -> Self;
    fn with_ymd(self, year: u16, month: u8, day: u8) -> Self;
    fn with_hour(self, hour: u8) -> Self;
    fn with_minute(self, minute: u8) -> Self;
    fn with_second(self, minute: u8) -> Self;
    fn with_hms(self, hour: u8, minute: u8, second: u8) -> Self;
}

impl TryFrom<time::OffsetDateTime> for DateTime {
    type Error = Box<dyn std::error::Error>;

    fn try_from(value: time::OffsetDateTime) -> Result<Self> {
        Ok(Self {
            year: value.year().try_into()?,
            month: value.month().into(),
            day: value.day(),
            hour: value.hour(),
            minute: value.minute(),
            second: value.second(),
        })
    }
}

impl TryFrom<time::PrimitiveDateTime> for DateTime {
    type Error = Box<dyn std::error::Error>;

    fn try_from(value: time::PrimitiveDateTime) -> Result<Self> {
        Ok(Self {
            year: value.year().try_into()?,
            month: value.month().into(),
            day: value.day(),
            hour: value.hour(),
            minute: value.minute(),
            second: value.second(),
        })
    }
}

impl TryFrom<time::Date> for DateTime {
    type Error = Box<dyn std::error::Error>;

    fn try_from(value: time::Date) -> Result<Self> {
        Ok(Self {
            year: value.year().try_into()?,
            month: value.month().into(),
            day: value.day(),
            hour: 0,
            minute: 0,
            second: 0,
        })
    }
}

impl TryFrom<DateTime> for PrimitiveDateTime {
    type Error = Box<dyn std::error::Error>;

    fn try_from(value: DateTime) -> std::result::Result<Self, Self::Error> {
        Ok(PrimitiveDateTime::new(
            time::Date::from_calendar_date(
                value.year.try_into()?,
                value.month.try_into()?,
                value.day,
            )?,
            time::Time::from_hms(value.hour, value.minute, value.second)?,
        ))
    }
}

impl TryFrom<DateTime> for OffsetDateTime {
    type Error = Box<dyn std::error::Error>;

    fn try_from(value: DateTime) -> std::result::Result<Self, Self::Error> {
        Ok(PrimitiveDateTime::try_from(value)?.assume_utc())
    }
}

impl PartialEq<OffsetDateTime> for DateTime {
    fn eq(&self, other: &OffsetDateTime) -> bool {
        let Ok(year) = i32::try_from(self.year) else {
            return false;
        };
        let Ok(month) = time::Month::try_from(self.month) else {
            return false;
        };
        year == other.year()
            && month == other.month()
            && self.day == other.day()
            && self.hour == other.hour()
            && self.minute == other.minute()
            && self.second == other.second()
    }
}

impl PartialEq<DateTime> for OffsetDateTime {
    fn eq(&self, other: &DateTime) -> bool {
        other.eq(self)
    }
}

impl FromStr for DateTime {
    type Err = Box<dyn std::error::Error>;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        let mut dt_parts = s.split('T');
        let date_part = dt_parts
            .next()
            .ok_or_else(|| format!("String too short: `{s}`"))?;
        let mut date_parts = date_part.split('-');
        let year: u16 = date_parts
            .next()
            .ok_or_else(|| format!("date part too short: `{s}`"))?
            .parse()?;
        let month: u8 = date_parts
            .next()
            .ok_or_else(|| format!("date part missing month: `{s}`"))?
            .parse()?;
        let day: u8 = date_parts
            .next()
            .ok_or_else(|| format!("date part missing day: `{s}`"))?
            .parse()?;
        let Some(time_part) = dt_parts.next() else {
            return Ok(Self {
                year,
                month,
                day,
                hour: 0,
                minute: 0,
                second: 0,
            });
        };
        let mut time_parts = time_part.split(":");
        let hour = time_parts
            .next()
            .and_then(|s| s.parse().ok())
            .unwrap_or_default();
        let minute = time_parts
            .next()
            .and_then(|s| s.parse().ok())
            .unwrap_or_default();
        let second = time_parts
            .next()
            .and_then(|s| {
                let end_idx = s
                    .find(|ch: char| !ch.is_ascii_digit())
                    .unwrap_or_else(|| s.len());
                s.get(..end_idx)?.parse().ok()
            })
            .unwrap_or_default();
        Ok(Self {
            year,
            month,
            day,
            hour,
            minute,
            second,
        })
    }
}

impl fmt::Display for DateTime {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("{:04}-", self.year))?;
        f.write_fmt(format_args!("{:02}-", self.month))?;
        f.write_fmt(format_args!("{:02}T", self.day))?;
        f.write_fmt(format_args!("{:02}:", self.hour))?;
        f.write_fmt(format_args!("{:02}:", self.minute))?;
        f.write_fmt(format_args!("{:02}.000Z", self.second))?;
        Ok(())
    }
}

impl Serialize for DateTime {
    fn serialize<S>(&self, s: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        s.serialize_str(&self.to_string())
    }
}

impl<'de> Deserialize<'de> for DateTime {
    fn deserialize<D>(d: D) -> std::result::Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        time::serde::rfc3339::deserialize(d)?
            .try_into()
            .map_err(|e| serde::de::Error::custom(e))
    }
}

impl PartialOrd for DateTime {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        use std::cmp::Ordering;
        Some(self.cmp(other))
    }
}

impl Ord for DateTime {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        use std::cmp::Ordering;
        fn try_prop<T>(l: T, r: T) -> Option<Ordering>
        where
            T: PartialOrd,
        {
            let ret = l.partial_cmp(&r)?;
            if matches!(ret, Ordering::Equal) {
                return None;
            }
            Some(ret)
        }
        if let Some(ret) = try_prop(self.year, other.year) {
            return ret;
        }
        if let Some(ret) = try_prop(self.month, other.month) {
            return ret;
        }
        if let Some(ret) = try_prop(self.day, other.day) {
            return ret;
        }
        if let Some(ret) = try_prop(self.hour, other.hour) {
            return ret;
        }
        if let Some(ret) = try_prop(self.minute, other.minute) {
            return ret;
        }
        if let Some(ret) = try_prop(self.second, other.second) {
            return ret;
        }
        Ordering::Equal
    }
}

#[cfg(test)]
mod tests {
    use rayon::iter::{IntoParallelIterator, ParallelIterator};
    use time::OffsetDateTime;

    use super::*;

    #[test]
    fn convert_revert() {
        do_convert_revert(65544);
    }

    #[test]
    fn convert_revert_many() {
        (0..=u16::MAX).into_iter().into_par_iter().for_each(|n| {
            do_convert_revert((n as i64) * 4);
        });
    }

    #[track_caller]
    fn do_convert_revert(timestamp: i64) {
        let dt = OffsetDateTime::from_unix_timestamp(timestamp).unwrap();
        let as_internal = DateTime::try_from(dt).unwrap();
        assert_eq!(dt, as_internal);
        let as_string = as_internal.to_string();
        let from_string: DateTime = as_string.parse().unwrap();
        assert_eq!(from_string, as_internal, "{from_string:?} != {as_string}");
        let from_internal = OffsetDateTime::try_from(from_string).unwrap();
        assert_eq!(dt, from_internal);
    }

    #[test]
    fn serde_round_trip() {
        do_serde_round_trip(66666);
    }

    #[test]
    fn serde_round_trip_many() {
        (0..=u16::MAX).into_iter().into_par_iter().for_each(|n| {
            do_serde_round_trip((n as i64) * 4);
        });
    }

    #[track_caller]
    fn do_serde_round_trip(timestamp: i64) {
        let dt = OffsetDateTime::from_unix_timestamp(timestamp).unwrap();
        let as_internal = DateTime::try_from(dt).unwrap();
        let json_string = serde_json::to_string(&as_internal).unwrap();
        let from_json: DateTime = serde_json::from_str(&json_string).unwrap();
        assert_eq!(as_internal, from_json);
    }
}
