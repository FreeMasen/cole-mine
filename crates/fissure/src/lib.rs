//! Database Abstractions
//! 

use std::{ops::RangeBounds, path::Path};

use date::DateTime;
use serde::{Deserialize, Serialize};
use structsy::{
    derive::queries,
    Filter, Operators, Structsy, StructsyTx,
};
use time::OffsetDateTime;
use crate::date::DateTimeQuery;

mod date;

type Result<T = (), E = Box<dyn std::error::Error>> = std::result::Result<T, E>;

#[derive(Clone)]
pub struct Database(Structsy);

impl Database {
    pub fn new(path: impl AsRef<Path>) -> Result<Self> {
        let inner =
            Structsy::open(path.as_ref()).map_err(|e| format!("Error opening database: {e}"))?;
        let ret = Self(inner);
        ret.init()?;
        Ok(ret)
    }

    #[cfg(test)]
    fn test() -> Result<Self> {
        let inner = Structsy::memory()?;
        let ret = Self(inner);
        ret.init()?;
        Ok(ret)
    }

    fn init(&self) -> Result {
        self.0.define::<Ring>()?;
        self.0.define::<RingEvent>()?;
        Ok(())
    }

    pub fn get_rings(&self) -> Vec<Ring> {
        self.0.query::<Ring>().into_iter().map(|(_, e)| e).collect()
    }

    pub fn get_ring(&self, mac: &str) -> Result<Ring> {
        let (_, ret) = self
            .0
            .query()
            .with_mac(mac)
            .fetch()
            .next()
            .ok_or_else(|| format!("unable to find ring with {mac}"))?;
        Ok(ret)
    }

    pub fn add_ring(&self, ring: &Ring) -> Result {
        let mut tx = self.0.begin()?;
        tx.insert(ring)?;
        tx.commit()?;
        Ok(())
    }

    pub fn update_ring(&self, ring: &Ring) -> Result {
        let mut tx = self.0.begin()?;
        let db = tx
            .query()
            .with_mac(&ring.mac)
            .fetch()
            .next()
            .ok_or_else(|| format!("unable to find ring with {}", ring.mac))?;
        tx.update(&db.0, ring)?;
        tx.commit()?;
        Ok(())
    }

    pub fn get_events_for_ring(&self, mac: &str, when: OffsetDateTime) -> Result<Vec<RingEvent>> {
        let min = when.date().midnight().assume_utc();
        let max = min
            .date()
            .next_day()
            .ok_or_else(|| format!("Missing next day {min}"))?
            .midnight()
            .assume_utc();
        let min = DateTime::try_from(min)?;
        let max = DateTime::try_from(max)?;
        let q = self
            .0
            .query::<RingEvent>()
            .with_ring_mac(mac)
            .and(|and| and.between_time(min..max));

        Ok(q.into_iter().map(|(_, event)| event).collect())
    }

    pub fn add_events(&self, events: &[RingEvent]) -> Result<()> {
        let mut tx = self.0.begin()?;

        for event in events {
            let existing = tx
                .query::<RingEvent>()
                .with_ring_mac(&event.mac)
                .and(|and| {
                    let filter = Filter::<DateTime>::new()
                        .with_ymd(event.when.year, event.when.month, event.when.day)
                        .with_hms(event.when.hour, event.when.minute, event.when.second);
                    and.with_when(filter)
                })
                .into_iter()
                .filter(|(_r, e)| {
                    std::mem::discriminant(&e.value) == std::mem::discriminant(&event.value)
                })
                .next();
            if let Some((r, _e)) = existing {
                println!("found matching event\n{event:?}\n{_e:?}");
                tx.update(&r, event)?;
            } else {
                tx.insert(event)?;
            }
        }
        tx.commit()?;
        Ok(())
    }
}

#[derive(Debug, structsy::derive::Persistent, Serialize, Deserialize, PartialEq)]
pub struct Ring {
    pub nickname: Option<String>,
    pub name: String,
    #[index(mode = "exclusive")]
    pub mac: String,
}

#[queries(Ring)]
trait FindRingByMac {
    // here is our condition method, to notice that the name of the parameter has to be exactly the same of the struct field.
    fn with_mac(self, mac: &str) -> Self;
}

#[derive(
    Debug,
    structsy::derive::Persistent,
    Serialize,
    Deserialize,
    PartialEq,
    bon::Builder,
)]
pub struct RingEvent {
    #[builder(into)]
    #[index(mode = "cluster")]
    pub mac: String,
    #[builder(into)]
    pub when: DateTime,
    pub value: EventData,
}

#[derive(Debug, structsy::derive::PersistentEmbedded, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", content = "data")]
pub enum EventData {
    HeartRate(u16),
    Sleep(u16),
    Stress(u16),
    Oxygen(u16),
    Activity(Activity),
}

impl EventData {
    pub fn activity(steps: u8, calories: f64, distance: u8) -> Self {
        EventData::Activity(Activity {
            steps,
            calories,
            distance,
        })
    }
    pub fn oxygen(value: u16) -> Self {
        EventData::Oxygen(value)
    }
    pub fn sleep(value: u16) -> Self {
        EventData::Sleep(value)
    }
    pub fn stress(value: u16) -> Self {
        EventData::Stress(value)
    }
    pub fn heart_rate(value: u16) -> Self {
        EventData::HeartRate(value)
    }
}

#[derive(Debug, structsy::derive::PersistentEmbedded, Serialize, Deserialize, PartialEq)]
pub struct Activity {
    pub steps: u8,
    pub calories: f64,
    pub distance: u8,
}

#[queries(RingEvent)]
trait FindEventByMac {
    fn with_ring_mac(self, mac: &str) -> Self;
    fn with_when(self, when: Filter<DateTime>) -> Self;
    fn between_time<R: RangeBounds<DateTime>>(self, when: R) -> Self;
}

#[cfg(test)]
mod tests {
    use std::{sync::atomic::AtomicUsize, time::Duration};

    use time::{Date, Month, Time};

    use super::*;

    static MAC: &str = "00:00:00:00:00:00";
    static MAC2: &str = "00:00:00:00:00:02";

    #[test]
    fn add_rings() {
        let db = Database::test().unwrap();
        let ring1 = Ring {
            mac: MAC.to_string(),
            nickname: None,
            name: "ring1".to_string(),
        };
        let ring2 = Ring {
            mac: MAC2.to_string(),
            nickname: None,
            name: "ring2".to_string(),
        };
        db.add_ring(&ring1).unwrap();
        db.add_ring(&ring2).unwrap();
        let from_db = db.get_rings();
        assert_eq!(from_db.len(), 2, "Invalid length of rings {from_db:?}");
        assert_eq!(from_db.as_slice(), [ring1, ring2].as_slice());
    }

    #[test]
    fn add_ring() {
        let db = Database::test().unwrap();
        let ring = Ring {
            mac: MAC.to_string(),
            nickname: None,
            name: "name".to_string(),
        };
        db.add_ring(&ring).unwrap();
        let from_db = db.get_ring(&ring.mac).unwrap();
        assert_eq!(from_db, ring);
    }

    #[test]
    fn serde_events() {
        let events = [
            RingEvent::builder()
                .mac(MAC)
                .when(DateTime::builder().year(2001).month(1).day(31).build())
                .value(EventData::activity(11, 222.0, 88))
                .build(),
            RingEvent::builder()
                .mac(MAC)
                .when(DateTime::builder().year(2001).month(1).day(31).build())
                .value(EventData::heart_rate(90))
                .build(),
            RingEvent::builder()
                .mac(MAC)
                .when(DateTime::builder().year(2001).month(1).day(31).build())
                .value(EventData::oxygen(11))
                .build(),
            RingEvent::builder()
                .mac(MAC)
                .when(DateTime::builder().year(2001).month(1).day(31).build())
                .value(EventData::Sleep(0))
                .build(),
            RingEvent::builder()
                .mac(MAC)
                .when(DateTime::builder().year(2001).month(1).day(31).build())
                .value(EventData::Stress(0))
                .build(),
        ];
        let json = serde_json::to_string_pretty(&events).unwrap();
        let back: Vec<RingEvent> = serde_json::from_str(&json).unwrap();
        assert_eq!(events.as_slice(), back.as_slice());
        insta::assert_snapshot!(json);
    }

    #[test]
    fn no_data_loss() {
        let db = Database::test().unwrap();
        let mut events = Vec::new();
        let mut time = OffsetDateTime::new_utc(
            Date::from_calendar_date(2001, time::Month::January, 31).unwrap(),
            Time::from_hms(0, 0, 0).unwrap(),
        );
        for i in 0..48 {
            events.push(RingEvent {
                mac: MAC.to_string(),
                when: time.try_into().unwrap(),
                value: super::EventData::Stress(i),
            });
            time += Duration::from_secs(60 * 60);
        }

        db.add_events(&events).unwrap();
        let from_db: Vec<_> =
            db.0.query::<RingEvent>()
                .fetch()
                .into_iter()
                .map(|(_, e)| e)
                .collect();
        assert_eq!(from_db, events)
    }

    #[test]
    fn time_search_works() {
        // const MAC: &str = "00:00:00:00:00:00";
        // let db = Database::test().unwrap();
        // let mut jan_events = Vec::new();
        // let mut feb_events = Vec::new();
        // let mut time = OffsetDateTime::new_utc(
        //     Date::from_calendar_date(2001, time::Month::January, 31).unwrap(),
        //     Time::from_hms(0, 0, 0).unwrap(),
        // );
        // let start = time;
        // while time.month() == Month::January {
        //     jan_events.push(RingEvent {
        //         mac: MAC.to_string(),
        //         when: DateTime(time),
        //         value: super::EventData::Stress(jan_events.len() as _),
        //     });
        //     time += Duration::from_secs(60 * 60);
        // }
        // for i in 0..24 {
        //     feb_events.push(RingEvent {
        //         mac: MAC.to_string(),
        //         when: DateTime(time),
        //         value: super::EventData::Stress(i),
        //     });
        //     time += Duration::from_secs(60 * 60);
        // }

        // db.add_events(&jan_events).unwrap();
        // db.add_events(&feb_events).unwrap();
        // let from_db = db.get_events_for_ring(MAC, start).unwrap();
        // assert_eq!(from_db, jan_events)
    }
}
