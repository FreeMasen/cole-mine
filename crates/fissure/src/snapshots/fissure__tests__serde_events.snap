---
source: crates/fissure/src/lib.rs
expression: json
---
[
  {
    "mac": "00:00:00:00:00:00",
    "when": "2001-01-31T00:00:00.000Z",
    "value": {
      "type": "Activity",
      "data": {
        "steps": 11,
        "calories": 222.0,
        "distance": 88
      }
    }
  },
  {
    "mac": "00:00:00:00:00:00",
    "when": "2001-01-31T00:00:00.000Z",
    "value": {
      "type": "HeartRate",
      "data": 90
    }
  },
  {
    "mac": "00:00:00:00:00:00",
    "when": "2001-01-31T00:00:00.000Z",
    "value": {
      "type": "Oxygen",
      "data": 11
    }
  },
  {
    "mac": "00:00:00:00:00:00",
    "when": "2001-01-31T00:00:00.000Z",
    "value": {
      "type": "Sleep",
      "data": 0
    }
  },
  {
    "mac": "00:00:00:00:00:00",
    "when": "2001-01-31T00:00:00.000Z",
    "value": {
      "type": "Stress",
      "data": 0
    }
  }
]
